bot/modules/dispute/scenes.ts

```
@@ -0,0 +1,64 @@
+import { Scenes } from 'telegraf';
+import { validateObjectId } from '../../validations';
+import { logger } from '../../../logger';
+import * as messages from './messages';
+import { CommunityContext } from '../community/communityContext';
+import { dispute } from './commands';
+
+export const DISPUTE_WIZARD = 'DISPUTE_WIZARD';
+
+export const middleware = () => {
+  const stage = new Scenes.Stage([disputeWizard]);
+  return stage.middleware();
+};
+
+export const disputeWizard = new Scenes.WizardScene(
+  DISPUTE_WIZARD,
+  async (ctx: CommunityContext) => {
+    try {
+      const { text } = messages.disputeWizardStatus(ctx.i18n, ctx.wizard.state);
+      const res = await ctx.reply(text);
+      ctx.wizard.state.currentStatusText = text;
+      ctx.wizard.state.statusMessage = res;
+      ctx.wizard.state.updateUI = async () => {
+        try {
+          const { text } = messages.disputeWizardStatus(ctx.i18n, ctx.wizard.state);
+          if (ctx.wizard.state.currentStatusText === text) return;
+          await ctx.telegram.editMessageText(
+            res.chat.id,
+            res.message_id,
+            undefined,
+            text
+          );
+          ctx.wizard.state.currentStatusText = text;
+        } catch (err) {
+          logger.error(err);
+        }
+      };
+      return ctx.wizard.next();
+    } catch (err) {
+      logger.error(err);
+      return ctx.scene.leave();
+    }
+  },
+  async (ctx) => {
+    try {
+      if (!ctx.message?.text) return;
+      const orderId = ctx.message.text.trim();
+      await ctx.deleteMessage();
+      if (!orderId || !(await validateObjectId(ctx, orderId))) {
+        await ctx.wizard.state.updateUI();
+        return;
+      }
+      ctx.user = ctx.wizard.state.user;
+      ctx.message.text = `/dispute ${orderId}`;
+      await dispute(ctx);
+      return ctx.scene.leave();
+    } catch (err) {
```

@pichardonuba why are you using so many catch blocks? are you expecting an exception in a non-exceptional case?

---

> This try/catch is necessary to ensure that if there’s an error inside the async function, the wizard doesn’t get stuck or crash.

Following that logic, every line of code in the project (or in any project) should be wrapped around a try-catch block.

---

I understand your point.
I added the try/catch specifically here because in the context of the wizard, if an unexpected error happens inside this async function, it could interrupt the flow of the scene and leave it stuck, affecting the user experience.
My intention isn’t to wrap every line of code, but just to protect the part that controls the scene flow, which I consider critical.
If you think there’s a better way to handle this, I’m open to suggestions.

---

@knocte I understand your point, so I’d like to explain more clearly why I decided to use a try/catch here:

1- It keeps consistency in the app’s structure: other wizards also use try/catch in the same part of the code to protect the flow.

2- It’s not exactly a silent failure: inside the catch we do logger.error(err); so the error gets logged, and we can also send a message to the user or leave the scene to prevent the wizard from getting stuck. I prefer to prioritize keeping the bot responsive even if something goes wrong.

3- In TypeScript it’s quite common to use try/catch when calling external APIs, databases, or file systems. In this case we’re using the Telegraf API, and I’d rather catch unexpected errors that could interrupt the bot’s flow.

If you have another suggestion to handle these errors while keeping both user experience and traceability, I’m totally open to it.

What do you think, @Catrya and @grunch ?